src/solvers/unsteady_solver.C

```diff
@@ -281,6 +294,168 @@ void UnsteadySolver::integrate_adjoint_sensitivity(const QoISet & qois, const Pa
 
 }
 
+void UnsteadySolver::integrate_adjoint_refinement_error_estimate(AdjointRefinementEstimator & adjoint_refinement_error_estimator, ErrorVector & QoI_elementwise_error, std::vector<Real *> QoI_time_instant)
+{
+  // Make sure the system::qoi_error_estimates vector is of the same size as system::qoi
+  if(_system.qoi_error_estimates.size() != _system.qoi.size())
+      _system.qoi_error_estimates.resize(_system.qoi.size());
+
+  // There are two possibilities regarding the integration rule we need to use for time integration.
+  // If we have a instantaneous QoI, then we need to use a left sided Riemann sum, otherwise the trapezoidal rule for temporally smooth QoIs.
+
+  // Create left and right error estimate vectors of the right size
+  std::vector<Number> qoi_error_estimates_left(_system.qoi.size());
+  std::vector<Number> qoi_error_estimates_right(_system.qoi.size());
+
+  // Get t_j
+  Real time_left = _system.time;
+
+  // Get f(t_j)
+  ErrorVector QoI_elementwise_error_left;
+  std::map<int, Real> global_spatial_errors_left;
+
+  // If we are at the very initial step, the error contribution is zero,
+  // otherwise the old ajoint vector has been filled and we are the left end
+  // of a subsequent timestep or sub-timestep
+  if((*old_adjoint).initialized())
+  {
+    // For evaluating the residual, we need to use the deltat that was used
+    // to get us to this solution, so we save the current deltat as next_step_deltat
+    // and set _system.deltat to the last completed deltat.
+    next_step_deltat = _system.deltat;
+    _system.deltat = last_step_deltat;
+
+    // The adjoint error estimate expression for a backwards facing step
+    // scheme needs the adjoint for the last time instant, so save the current adjoint for future use
+    NumericVector<Number> & current_adjoint = _system.get_adjoint_solution(0);
```

Comment:
This only works if we have a single QoI at index 0?  That needs to be fixed.  Maybe with a QoISet argument to this function?

Reply:
Ah, sorry, didn't see that you're using the QoISet attached to the ARefEE.

Reply:
But we should still be doing that same loop here, no?

Reply:
You are right, this needs to be generalized.